code/harry_potter_store/src/store.py

```diff
@@ -0,0 +1,85 @@
+from typing import Dict
+
+
+class Store:
+    book_types = ["1", "2", "3", "4", "5"]
+
+    def __init__(self, book_price=8):
+        self.book_price = book_price
+        pass
+
+    def _validate(self, to_buy_books: Dict[str, int]):
+        for book_type, amount in to_buy_books.items():
+            if book_type not in Store.book_types:
+                raise ValueError("Book type %s must be in %s".format(book_type, Store.book_types))
+            if amount <= 0:
+                raise ValueError("Book amount %s for type %s must be > 0".format(amount, book_type))
+
+    def _clean(self, to_buy_books: Dict[str, int]) -> dict:
+        return dict(filter(lambda item: item[1] > 0 and item[0] in Store.book_types, to_buy_books.items()))
+
+    def _decrement(self, to_buy_books: Dict[str, int], count: int) -> dict:
+        first_part = {book: amount - 1 for book, amount in list(to_buy_books.items())[:count]}
+        second_part = dict(list(to_buy_books.items())[count:])
+        first_part.update(second_part)
+        return self._clean(first_part)
+
+    def _discount_wrapper(self, number, discount):
+        return lambda to_buy_books, price, predicate: \
+            self._discount(to_buy_books, price, predicate, number, discount)
+
+    def _discount(self, to_buy_books, price, predicate, number, discount):
+        if predicate(len(to_buy_books), number):
+            price += number * self.book_price * (1 - discount)
+            to_buy_books = self._decrement(to_buy_books, number)
+        return to_buy_books, price
+
+    """
+    algorithm logic: we have discounts for several books. We give names for it: 3, 2, 1
+    and we are doing a full sort out
+    we calculate a price just with discount for 3 books
+    next for 2 books
+    next for 1 books
+    and next we calculate a price with discount for 3, 2, 1 books
+    next for 2, 1 books
+    """
+    def get_price(self, to_buy_books: Dict[str, int]) -> float:
+        """
+        to_buy_books - dictionary from book type and book amount
+        key of dict -- book type
+        value of dict -- book amount
+        """
+        self._validate(to_buy_books)
+        to_buy_books = {k: v for k, v in sorted(to_buy_books.items(), key=lambda item: item[1], reverse=True)}
+        discounts = [self._discount_wrapper(5, 0.25), self._discount_wrapper(4, 0.2),
+                     self._discount_wrapper(3, 0.1), self._discount_wrapper(2, 0.05)]
+        min_price = self.book_price * sum(to_buy_books.values())
+        for i in range(len(discounts)):
+            min_price = min(min_price,
+                            self._get_price(to_buy_books,
+                                            [discounts[i]], len(discounts) - i, lambda x, y: x >= y))
+            if i != 0:
+                min_price = min(min_price,
+                                self._get_price(to_buy_books,
+                                                discounts[0:i + 1], len(discounts) - i, lambda x, y: x == y))
```

Comment:
```suggestion
        min_price = min(min_price, 
                                   self._get_price(to_buy_books, discounts, 1, lambda x, y: x == y))
```

Reply:
i don't understand what you mean

Reply:
This is code suggestion according to my previous comments. How can it be more clear than code change? Could you please read previous comments and compare your code with the suggestion?

Reply:
so, why we can not missed case with best discount 3,2 for example? 
if a lot of discount will be apply by greedy strategy and we missed another best set of discounts? 
i change just greedy strategy to full search

What strategy you suggest?

Reply:
If you apply current code suggestion, all tests will be still passing. If you're sure there will be some miss then please add corresponding test case which would fail :)

From comment above:
>Use multiple discounts in a greedy way with increasing subsets of discounts like [5,4], [5,4,3], [5,4,3,2] (seems similar to the previous algorithm)

It looks obvious that greedy algorithm using just [5,4,3,2] will produce the same result as running greedy algorithm 3 times with  [5,4], [5,4,3], [5,4,3,2] (by definition greedy takes the biggest available set of discounts)

Reply:
yep, i changed sequence of discount functions from [5, 4], [5, 4, 3]... to [3, 2], [4, 3 ,2]...